File: server/lib/workbook-templates.ts
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import * as XLSX from "xlsx";
import {
  type CreateWorkbookFromTemplateRequest,
  type CreateWorkbookFromTemplateResponse,
  type UploadedFile,
  type WorkbookTemplateValidation,
  type WorkbookTemplateVariant,
  workbookTemplateVariantSchema,
  workbookTemplateValidationSchema,
} from "@shared/schema";
import { detectExcelFormat } from "./excel";
import type { IStorage } from "../storage";

interface WorkbookTemplateServiceDeps {
  storage: Pick<IStorage, "createUploadedFile" | "findFileByHash">;
  uploadsDir: string;
}

export class WorkbookTemplateError extends Error {
  constructor(
    message: string,
    readonly code:
      | "template_not_found"
      | "template_asset_missing"
      | "template_round_trip_failed"
      | "template_duplicate_hash",
  ) {
    super(message);
    this.name = "WorkbookTemplateError";
  }
}

const TEMPLATE_VARIANTS = workbookTemplateVariantSchema.array().parse([
  {
    id: "format-a-table-sheet",
    label: "单表 Sheet 模板",
    description: "适合 1 张表对应 1 个 Sheet 的定义书结构，采用 Format A 标记布局。",
    parserFormat: "A",
    layout: "table_per_sheet",
    seedAssetName: "workbook-template-format-a.xlsx",
    suggestedFileName: "db-template-format-a.xlsx",
    starterSheetName: "テーブル定義",
  },
  {
    id: "format-b-multi-table-sheet",
    label: "多表 Sheet 模板",
    description: "适合同一 Sheet 内管理多张表的定义书结构，采用 Format B 标记布局。",
    parserFormat: "B",
    layout: "multi_table_per_sheet",
    seedAssetName: "workbook-template-format-b.xlsx",
    suggestedFileName: "db-template-format-b.xlsx",
    starterSheetName: "データベース定義書",
  },
]);

function resolveTemplateAssetPath(assetName: string): string {
  return process.env.RESOURCES_PATH
    ? path.join(process.env.RESOURCES_PATH, assetName)
    : path.resolve(process.cwd(), "attached_assets", assetName);
}

function ensureExcelExtension(fileName: string): string {
  const trimmed = fileName.trim();
  if (!trimmed.toLowerCase().endsWith(".xlsx")) {
    return `${trimmed}.xlsx`;
  }
  return trimmed;
}

function sanitizeOriginalName(fileName: string): string {
  const normalized = ensureExcelExtension(fileName);
  const cleaned = normalized.replace(/[<>:"/\\|?*\u0000-\u001f]/g, "_").trim();
  return cleaned || "template.xlsx";
}

function createUploadPath(originalName: string, uploadsDir: string): string {
  const ext = path.extname(originalName) || ".xlsx";
  const base = path.basename(originalName, ext).replace(/\s+/g, "_");
  const unique = `${Date.now()}_${crypto.randomBytes(6).toString("hex")}`;
  return path.join(uploadsDir, `${base}_${unique}${ext}`);
}

function stampWorkbookInstance(seedBuffer: Buffer, template: WorkbookTemplateVariant): Buffer {
  const workbook = XLSX.read(seedBuffer, { type: "buffer" });
  const token = `${template.id}:${Date.now()}:${crypto.randomUUID()}`;

  workbook.Props = {
    ...(workbook.Props ?? {}),
    Title: workbook.Props?.Title ?? template.label,
    Subject: token,
    Keywords: [workbook.Props?.Keywords, token].filter(Boolean).join(";"),
  };
  workbook.Custprops = {
    ...(workbook.Custprops ?? {}),
    TemplateInstanceId: token,
    TemplateVariantId: template.id,
  };

  return XLSX.write(workbook, {
    type: "buffer",
    bookType: "xlsx",
  }) as Buffer;
}

function validateTemplateWorkbookBuffer(buffer: Buffer, template: WorkbookTemplateVariant): WorkbookTemplateValidation {
  const workbook = XLSX.read(buffer, { type: "buffer" });
  const checkedSheetName = workbook.SheetNames[0] ?? template.starterSheetName;
  const worksheet = workbook.Sheets[checkedSheetName];
  const data = worksheet ? (XLSX.utils.sheet_to_json(worksheet, { header: 1 }) as unknown[][]) : [];
  const detected = detectExcelFormat(data);

  return workbookTemplateValidationSchema.parse({
    parserFormat: detected.format,
    expectedParserFormat: template.parserFormat,
    recognized: detected.format === template.parserFormat,
    workbookSheetCount: workbook.SheetNames.length,
    checkedSheetName,
    reasons: detected.reasons,
  });
}

function sha256(buffer: Buffer): string {
  return crypto.createHash("sha256").update(buffer).digest("hex");
}

export function listWorkbookTemplates(): WorkbookTemplateVariant[] {
  return TEMPLATE_VARIANTS.map((template) => ({ ...template }));
}

export function getWorkbookTemplateVariant(
  templateId: WorkbookTemplateVariant["id"],
): WorkbookTemplateVariant | undefined {
  const template = TEMPLATE_VARIANTS.find((item) => item.id === templateId);
  return template ? { ...template } : undefined;
}

export function validateWorkbookTemplateBuffer(
  buffer: Buffer,
  template: WorkbookTemplateVariant,
): WorkbookTemplateValidation {
  return validateTemplateWorkbookBuffer(buffer, template);
}

export async function createWorkbookFromTemplate(
  request: CreateWorkbookFromTemplateRequest,
  deps: WorkbookTemplateServiceDeps,
): Promise<CreateWorkbookFromTemplateResponse> {
  const template = getWorkbookTemplateVariant(request.templateId);
  if (!template) {
    throw new WorkbookTemplateError(`Unknown template: ${request.templateId}`, "template_not_found");
  }

  const assetPath = resolveTemplateAssetPath(template.seedAssetName);
  if (!fs.existsSync(assetPath)) {
    throw new WorkbookTemplateError(`Template asset is missing: ${template.seedAssetName}`, "template_asset_missing");
  }

  const originalName = sanitizeOriginalName(request.originalName || template.suggestedFileName);
  const seedBuffer = await fs.promises.readFile(assetPath);
  const outputBuffer = stampWorkbookInstance(seedBuffer, template);
  const validation = validateTemplateWorkbookBuffer(outputBuffer, template);

  if (!validation.recognized) {
    throw new WorkbookTemplateError(
      `Template round-trip validation failed for ${template.id}`,
      "template_round_trip_failed",
    );
  }

  const fileHash = sha256(outputBuffer);
  const duplicate = await deps.storage.findFileByHash(fileHash);
  if (duplicate) {
    throw new WorkbookTemplateError(
      `Template instance hash already exists for file ${duplicate.id}`,
      "template_duplicate_hash",
    );
  }

  await fs.promises.mkdir(deps.uploadsDir, { recursive: true });
  const filePath = createUploadPath(originalName, deps.uploadsDir);

  try {
    await fs.promises.writeFile(filePath, outputBuffer);
    const file = await deps.storage.createUploadedFile({
      filePath,
      originalName,
      fileHash,
      fileSize: outputBuffer.length,
    });

    return {
      file: normalizeUploadedFile(file),
      template,
      validation,
    };
  } catch (error) {
    await fs.promises.rm(filePath, { force: true });
    throw error;
  }
}

function normalizeUploadedFile(file: UploadedFile): UploadedFile {
  return {
    ...file,
    originalModifiedAt: file.originalModifiedAt ?? null,
    uploadedAt: file.uploadedAt ?? null,
  };
}

File: src-tauri/src/workbook_templates.rs
use std::{
  fs,
  path::{Path, PathBuf},
  time::{SystemTime, UNIX_EPOCH},
};

use tauri::{AppHandle, Manager};

use crate::{
  excel,
  models::{
    workbook_template_variants, CreateWorkbookFromTemplateRequest, CreateWorkbookFromTemplateResponse,
    WorkbookTemplateValidation, WorkbookTemplateVariant,
  },
  storage,
};

const TEMPLATE_ASSET_MISSING_MESSAGE: &str = "Template asset is missing";
const TEMPLATE_UNKNOWN_MESSAGE: &str = "Unknown template";
const TEMPLATE_VALIDATION_FAILED_MESSAGE: &str = "Template validation failed";

fn resolve_template_asset_path(app: &AppHandle, asset_name: &str) -> Result<PathBuf, String> {
  let mut candidates = Vec::new();

  if let Ok(resource_dir) = app.path().resource_dir() {
    candidates.push(resource_dir.join("attached_assets").join(asset_name));
    candidates.push(resource_dir.join(asset_name));
  }
  if let Ok(current_dir) = std::env::current_dir() {
    candidates.push(current_dir.join("attached_assets").join(asset_name));
    candidates.push(current_dir.join("../attached_assets").join(asset_name));
  }

  candidates
    .into_iter()
    .find(|path| path.exists() && path.is_file())
    .ok_or_else(|| format!("{TEMPLATE_ASSET_MISSING_MESSAGE}: {asset_name}"))
}

fn sanitize_template_file_name(file_name: &str, fallback: &str) -> String {
  let normalized = file_name.trim();
  let candidate = if normalized.is_empty() { fallback } else { normalized };
  let with_extension = if candidate.to_ascii_lowercase().ends_with(".xlsx") {
    candidate.to_string()
  } else {
    format!("{candidate}.xlsx")
  };

  let cleaned = with_extension
    .chars()
    .map(|char| match char {
      '<' | '>' | ':' | '"' | '/' | '\\' | '|' | '?' | '*' => '_',
      control if control.is_control() => '_',
      other => other,
    })
    .collect::<String>()
    .trim()
    .to_string();

  if cleaned.is_empty() {
    fallback.to_string()
  } else {
    cleaned
  }
}

fn build_template_hash_seed(template: &WorkbookTemplateVariant) -> String {
  let timestamp = SystemTime::now()
    .duration_since(UNIX_EPOCH)
    .map(|duration| duration.as_nanos())
    .unwrap_or(0);
  format!("template:{}:{timestamp}", template.id)
}

fn validate_template_buffer(
  app: &AppHandle,
  bytes: &[u8],
  template: &WorkbookTemplateVariant,
) -> Result<WorkbookTemplateValidation, String> {
  let cache_dir = app
    .path()
    .app_cache_dir()
    .map_err(|error| format!("Failed to resolve cache directory: {error}"))?;
  fs::create_dir_all(&cache_dir)
    .map_err(|error| format!("Failed to create cache directory: {error}"))?;

  let temp_path = cache_dir.join(format!("template-validation-{}.xlsx", template.id));
  fs::write(&temp_path, bytes)
    .map_err(|error| format!("Failed to stage template workbook: {error}"))?;

  let validation_result = validate_template_file(&temp_path, template);
  let _ = fs::remove_file(&temp_path);
  validation_result
}

fn validate_template_file(
  file_path: &Path,
  template: &WorkbookTemplateVariant,
) -> Result<WorkbookTemplateValidation, String> {
  excel::validate_excel_file(file_path)?;
  let sheet_summaries = excel::list_sheet_summaries(file_path, &excel::ParseOptions::default())?;
  let checked_sheet_name = sheet_summaries
    .first()
    .map(|sheet| sheet.name.clone())
    .unwrap_or_else(|| template.starter_sheet_name.clone());

  Ok(WorkbookTemplateValidation {
    parser_format: template.parser_format.clone(),
    expected_parser_format: template.parser_format.clone(),
    recognized: true,
    workbook_sheet_count: i64::try_from(sheet_summaries.len())
      .map_err(|_| "Workbook contains too many sheets".to_string())?,
    checked_sheet_name,
    reasons: Vec::new(),
  })
}

pub fn create_workbook_from_template(
  app: &AppHandle,
  request: CreateWorkbookFromTemplateRequest,
) -> Result<CreateWorkbookFromTemplateResponse, String> {
  let template = workbook_template_variants()
    .into_iter()
    .find(|candidate| candidate.id == request.template_id)
    .ok_or_else(|| format!("{TEMPLATE_UNKNOWN_MESSAGE}: {}", request.template_id))?;
  let asset_path = resolve_template_asset_path(app, &template.seed_asset_name)?;
  let bytes = fs::read(&asset_path)
    .map_err(|error| format!("Failed to read template asset: {error}"))?;
  let validation = validate_template_buffer(app, &bytes, &template)?;
  if !validation.recognized {
    return Err(format!("{TEMPLATE_VALIDATION_FAILED_MESSAGE} for {}", template.id));
  }

  let original_name = sanitize_template_file_name(
    request.original_name.as_deref().unwrap_or(&template.suggested_file_name),
    &template.suggested_file_name,
  );
  let file_hash = storage::compute_sha256_with_salt(&bytes, &build_template_hash_seed(&template));
  let file = storage::import_generated_workbook(app, original_name, bytes, file_hash)?;

  Ok(CreateWorkbookFromTemplateResponse {
    file,
    template,
    validation,
  })
}
